[FEAT] read_sql

colin-ho opened the pull request:
Closes #1560 

Adds a new read method: `read_sql(sql: str, url: str)`, which executes a given sql query on a given database url, and creates a Dataframe from the results.

Drive bys:
- Added a from_arrow cast from pyarrow date64 to daft timestamp (both are millisecond precision) to allow connectorx reads with timestamp columns.

Features:
- Uses [connector-x](https://github.com/sfu-db/connector-x/tree/main), which reads directly into pyarrow via rust when supported, else fallback to [SQL alchemy ](https://docs.sqlalchemy.org/en/20/orm/quickstart.html)
- Partitioned reads via limit and offset when supported
- Integration tests for Postgres, MySQL, Trino
- Pushdowns into base SQL query

colin-ho commented on daft/sql/sql_reader.py:
```diff
@@ -0,0 +1,106 @@
+
+        if self.limit is not None and self.offset is not None:
+            if self.apply_limit_before_offset is True:
+                clauses.append(f"ORDER BY 1 LIMIT {self.limit} OFFSET {self.offset}")
```
Added order by 1 because Trino arbitrarily discards rows without a specified order by clause.  https://trino.io/docs/current/sql/select.html#offset-clause

colin-ho replied:
Actually, order by will introduce unnecessary sorting, so this might not be a good choice. I'll investigate a bit more on how the trino queries can be partitioned, worst case we don't do partitioned reads for Trino

colin-ho replied:
Actually this isn't just a Trino thing: https://www.postgresql.org/docs/current/queries-limit.html ```using different LIMIT/OFFSET values to select different subsets of a query result will give inconsistent results unless you enforce a predictable result ordering with ORDER BY. This is not a bug; it is an inherent consequence of the fact that SQL does not promise to deliver the results of a query in any particular order unless ORDER BY is used to constrain the order.``` 

so i think if we keep the limit and offsets we have no choice but to do order by, but then theres also potential issues if the order by column is not unique

samster25 commented on src/daft-dsl/src/expr.rs:
```diff
@@ -568,6 +569,86 @@ impl Expr {
+                    to_sql_inner(right, buffer)
+                }
+                Expr::Not(inner) => {
+                    write!(buffer, "NOT (")?;
```
you should be able to collapse this 
`write!(buffer, "NOT ({})", o_sql_inner(inner, buffer)?)`

samster25 commented on src/daft-dsl/src/expr.rs:
```diff
@@ -568,6 +569,86 @@ impl Expr {
+                    if_false,
+                    predicate,
+                } => {
+                    write!(buffer, "CASE WHEN ")?;
```
this should be able to be a single `write!` macro

samster25 commented on src/daft-dsl/src/lit.rs:
```diff
@@ -212,6 +212,36 @@ impl LiteralValue {
         result
     }
+
+    pub fn to_sql(&self) -> Option<String> {
```
no need to implement but you could have this be `display_sql` and take in a formatter as well!

samster25 commented on src/daft-scan/src/python.rs:
```diff
@@ -319,6 +319,35 @@ pub mod pylib {
+                size_bytes,
+                metadata: num_rows.map(|n| TableMetadata { length: n as usize }),
+                partition_spec: None,
+                statistics: None,
```
if the table is partitioned by some column, we could leverage that for statistics

colin-ho replied:
Will make an issue for this to do as a follow on.

samster25 reviewed (commented):
Nice work! Just some nits :)

colin-ho commented on daft/table/table_io.py:
```diff
@@ -212,6 +213,41 @@ def read_parquet(
+        url,
+        # TODO(Colin): Enable pushdowns
+        projection=None,
+        predicate=None,
```
I left it out so that we could enable the predicate pushdown into SQL query in a following PR as a perf improvement. But I'm also good to include them in this one, the tests should all still pass!

colin-ho commented on daft/sql/sql_reader.py:
```diff
@@ -0,0 +1,93 @@
+                rows = result.fetchall()
+                pydict = {column_name: [row[i] for row in rows] for i, column_name in enumerate(result.keys())}
+
+                return pa.Table.from_pydict(pydict)
```
Made an issue to create type mappings from dbapi type codes

colin-ho commented:
> Nice work! Just some nits :)

Thanks! Addressed your feedback in latest commit. I also added functionality to insert limit pushdowns into SQL, as well as some general refactoring.

samster25 reviewed (approved):
🚀
